File: javascript/house/house.js
import {from} from './iterable-range';

const actors = [
  { who: 'house that Jack built.' },
  { who: 'malt', what: 'lay in' },
  { who: 'rat', what: 'ate' },
  { who: 'cat', what: 'killed' },
  { who: 'dog', what: 'worried' },
  { who: 'cow with the crumpled horn', what: 'tossed' },
  { who: 'maiden all forlorn', what: 'milked' },
  { who: 'man all tattered and torn', what: 'kissed' },
  { who: 'priest all shaven and shorn', what: 'married' },
  { who: 'rooster that crowed in the morn', what: 'woke' },
  { who: 'farmer sowing his corn', what: 'kept' },
  { who: 'horse and the hound and the horn', what: 'belonged to' },
];

const verse = (n) => {
  const lines = [`This is the ${actors[n - 1].who}`];
  for (let i = n - 1; i >= 1; i -= 1) {
    lines.push(`that ${actors[i].what} the ${actors[i - 1].who}`);
  }
  return lines;
};

const verses = (m = 1, n = actors.length) => {
  const lyrics = from(m)
                 .upTo(n)
                 .reduce((lyrics, i) => lyrics.concat(verse(i), ''), []);
  lyrics.pop();
  return lyrics;
};

module.exports = {House: {verse, verses}};

File: lua/house/house.lua
local house = {}

local items = {
    {the = "house", that = "Jack built."},
    {the = "malt",  that = "lay in"},
    {the = "rat",   that = "ate"},
    {the = "cat",   that = "killed"},
    {the = "dog",   that = "worried"},
    {the = "cow with the crumpled horn",       that = "tossed"},
    {the = "maiden all forlorn",               that = "milked"},
    {the = "man all tattered and torn",        that = "kissed"},
    {the = "priest all shaven and shorn",      that = "married"},
    {the = "rooster that crowed in the morn",  that = "woke"},
    {the = "farmer sowing his corn",           that = "kept"},
    {the = "horse and the hound and the horn", that = "belonged to"}
}

function house.verse(n)
    local verse = "This is"
    for i = n, 1, -1 do
        local sep = i == 1 and " " or "\n"
        verse = verse .. " the " .. items[i].the .. sep .. "that " .. items[i].that
    end
    return verse
end

function house.recite()
    local verses = {}
    for i = 1, #items do
        table.insert(verses, house.verse(i))
    end
    return table.concat(verses, "\n")
end

return house
